### PR TITLE
feat(dyn-sampling): Add backend validation for custom rules  [TET-148]

### PR DESCRIPTION
**Pull request by andriisoldatenko**
When adding an individual transaction rule in sampling,
users can add "Custom Tags" (key-value pairs),
but we would like to restrict some keys as they are used by Sentry.

Check that keys are not in:

https://github.com/getsentry/sentry/blob/2e5e658a56805c74a25eed88b8c59a488d9c652e/src/sentry/constants.py#L599-L626

```
➜  ~ curl 'http://localhost:8000/api/0/projects/sentry/earth/' \
  -X 'PUT' \
  -H 'Accept-Language: en-GB,en-US;q=0.9,en;q=0.8' \
  -H 'Connection: keep-alive' \
  -H 'Cookie: ****' \
  -H 'Origin: http://localhost:8000' \
  -H 'Referer: http://localhost:8000/settings/sentry/projects/earth/server-side-sampling/' \
  -H 'accept: application/json; charset=utf-8' \
  -H 'content-type: application/json' \
  -H 'x-csrftoken: ***' \
  --data-raw '{"dynamicSampling":{"rules":[{"id":0,"active":false,"type":"trace","condition":{"op":"and","inner":[{"op":"eq","name":"browser","value":["test"],"options":{"ignoreCase":true}}]},"sampleRate":0.12},{"sampleRate":1,"type":"trace","condition":{"op":"and","inner":[]},"active":true,"id":3}]}}' \
  --compressed
```

```
{"dynamicSampling":{"non_field_errors":["browser is not allowed as tag name"]}}%
```

### CODE REVIEW

File: src/sentry/api/endpoints/project_details.py [hunk 1]
```diff
@@ -106,6 +106,14 @@ def _is_uniform_sampling_rule(rule):
         # future, we will have to update this to also support single transactions.
         return len(rule["condition"]["inner"]) == 0 and rule["type"] == "trace"
 
+    def validate_custom_rules_tag_names(self, rules):
+        # Ensure that all custom tags are valid name
+        # and are not presented in DS_DENYLIST
+        for rule in rules:
+            for op in rule["condition"]["inner"]:
+                if op["name"] in DS_DENYLIST:
```
Comment (wedamija) [inline src/sentry/api/endpoints/project_details.py:114]: Is `name` always present here?

**Review by wedamija — APPROVED** [review]
Do we need to check that there aren't existing rules like this that use invalid ops?

File: src/sentry/api/endpoints/project_details.py (hunk 1 above)
Comment (andriisoldatenko) [reply src/sentry/api/endpoints/project_details.py:114]: @wedamija yeah i had similar thoughts, so for security let me add extra check.

@matejminar can you please help to figure out if `name` can be omitted.

Comment (andriisoldatenko) [reply src/sentry/api/endpoints/project_details.py]: fixed in 640b796f16

Comment (matejminar) [reply src/sentry/api/endpoints/project_details.py]: @andriisoldatenko frontend is always sending the name. what happens if someone fires the request without the name manually?

**Review by matejminar — APPROVED** [review]

**Review by priscilawebdev — APPROVED** [review]
lgtm

**Comment by andriisoldatenko** [comment]
we decided to postpone merging.

@jan-auer suggested
> we want to avoid pitfalls for users when we suggest them sampling by a tag but in fact it won't work because the tag isn't > ingested. i'd separate concerns here:

>   - the API runs strict validation.

>   - the UI provides additional guidance.

> this deny list is guidance

**Pull request closed without merging**